miri-script/src/commands.rs

```diff
@@ -0,0 +1,652 @@
+            Subcommands::Fmt { flags } => self.fmt(flags),
+            Subcommands::Clippy { flags } => self.clippy(flags),
+            Subcommands::Cargo { flags } => self.cargo(flags),
+            Subcommands::ManySeeds { command, seed_start, seeds } =>
+                self.many_seeds(command, *seed_start, *seeds),
+            Subcommands::Bench { benches } => self.bench(benches),
+            Subcommands::Toolchain { flags } => self.toolchain(flags),
+            Subcommands::RustcPull { commit } => self.rustc_pull(commit.clone()),
+            Subcommands::RustcPush { github_user, branch } => self.rustc_push(github_user, branch),
+        }
+    }
+
+    fn set_env(
+        &mut self,
+        key: impl Into<OsString>,
+        value: impl Into<OsString>,
+    ) -> Option<OsString> {
+        self.env.insert(key.into(), value.into())
+    }
+
+    /// Prepare and set MIRI_SYSROOT. Respects `MIRI_TEST_TARGET` and takes into account
+    /// locally built vs. distributed rustc.
+    fn find_sysroot(&mut self) -> Result<()> {
+        let current_sysroot = std::env::var_os("MIRI_SYSROOT").unwrap_or_default();
+
+        if !current_sysroot.is_empty() {
+            // Sysroot already set, use that.
+            let current_value = self.set_env("MIRI_SYSROOT", &current_sysroot);
+            assert!(current_value.is_none() || current_value.unwrap() == current_sysroot);
+            return Ok(());
+        }
+        // We need to build a sysroot.
+        let target = std::env::var_os("MIRI_TEST_TARGET").filter(|target| !target.is_empty());
+        let sysroot = self.build_sysroot(target.as_deref())?;
+        self.set_env("MIRI_SYSROOT", sysroot);
+        Ok(())
+    }
+
+    /// Build a sysroot and set MIRI_SYSROOT to use it. Arguments are passed to `cargo miri setup`.
+    fn build_sysroot(&self, target: Option<&OsStr>) -> Result<String> {
+        let manifest_path = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        let Self { active_toolchain, cargo_extra_flags, .. } = &self;
+        let target_prefix: Option<&OsStr> = target.map(|_| "--target".as_ref());
+        let sh = self.shell_with_miri_env()?;
+        let output = cmd!(sh, "cargo +{active_toolchain} run {cargo_extra_flags...} --manifest-path {manifest_path} -- miri setup --print-sysroot {target_prefix...} {target...}").read();
+        if output.is_err() {
+            // Run it again so the user can see the error.
+            // This time without printing a sysroot though.
+            let _ = cmd!(sh, "cargo +{active_toolchain} run {cargo_extra_flags...} --manifest-path {manifest_path} -- miri setup {target_prefix...} {target...}").run().with_context(|| "`cargo miri setup` failed")?;
+        }
+
+        Ok(output?)
+    }
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> OsString {
+        format!("+{}", &self.active_toolchain).into()
+    }
```

Comment:
This doesn't seem worth it -- `build_sysroot` looks just fine without this helper?

If we want a helper function it would be for invoking `cargo +toolchain <subcommand>`, but it seems with xshell that isn't really needed any more.